contrib/generate-struct.py

```diff
@@ -0,0 +1,381 @@
+            tmp_type = "guint32"
+        elif self.type == Type.U32:
+            tmp_type = "guint32"
+        elif self.type == Type.U64:
+            tmp_type = "guint64"
+        elif self.type == Type.STRING:
+            tmp_type = "gchar"
+        if self.multiplier:
+            tmp_type += "*"
+        return tmp_type
+
+    @property
+    def type_glib_const(self) -> str:
+        if self.multiplier:
+            return "const " + self.type_glib_mut
+        return self.type_glib_mut
+
+    def parse_default(self, val: str) -> None:
+
+        val = val.replace("$struct_size", str(self.size))
+        self.default = val
+
+    def parse_constant(self, val: str) -> None:
+
+        val = val.replace("$struct_size", str(self.size))
+        self.constant = val
+
+    def parse_type(self, val: str) -> None:
+
+        # get the integer multiplier
+        split_index = 0
+        for char in val:
+            if not char.isnumeric():
+                break
+            split_index += 1
+        if split_index:
+            self.multiplier = int(val[:split_index])
+
+        # find the type
+        try:
+            val2 = val[split_index:]
+            if val2.endswith("be"):
+                self.endian = Endian.BIG
+                val2 = val2[:-2]
+            elif val2.endswith("le"):
+                self.endian = Endian.LITTLE
+                val2 = val2[:-2]
+            self.type = Type(val2)
+        except ValueError as e:
+            raise ValueError(f"invalid type: {val2}") from e
+
+    def __str__(self) -> str:
+        tmp = f"{self.element_id}: "
+        if self.multiplier:
+            tmp += str(self.multiplier)
+        tmp += str(self.type)
+        if self.endian != Endian.NATIVE:
+            tmp += str(self.endian)
+        if self.default:
+            tmp += f": {self.default}"
+        elif self.constant:
+            tmp += f":: {self.constant}"
+        return tmp
+
+
+# convert a CamelCase name into snake_case
+def _camel_to_snake(name: str) -> str:
+
+    name_snake: str = ""
+    for char in name:
+        if char.islower():
+            name_snake += char
+            continue
+        if name_snake:
+            name_snake += "_"
+        name_snake += char.lower()
+    return name_snake
+
+
+class Generator:
+    def __init__(self, basename):
+        self.basename: str = basename
+
+    def _process_items(self, name: str, items: List[Item]) -> Tuple[str, str]:
+
+        # useful constants
+        name_snake = _camel_to_snake(name)
+
+        # header
+        str_h = "/* auto-generated, do not modify */\n"
+        str_h += "#include <glib.h>\n"
+        str_h += f"GByteArray* {name_snake}_new(void);\n"
+        str_h += f"GByteArray* {name_snake}_unpack(const guint8 *buf, gsize bufsz, gsize offset, GError **error);\n"
+        for item in items:
+            if item.element_id.startswith("_"):
+                continue
+            if item.constant:
+                continue
```

Comment:
really all of these should log something about how it's not conformant, otherwise we're going to see weird bugs

Reply:
aha, so the logic here is that the user never needs getters or setters for something that's a constant. It's either set by `_new()` or `_parse()` fails to run.

Reply:
I've added some docs